### PR TITLE
Added some additional exceptions

### PR DESCRIPTION
**Pull request by danielstokes**
This doesn't fix "\n" being parsed or even empty values.  It throws exceptions when preg_replace fails or does nothing.  I'm not sure if you'd like additional exception types added.  Just provide some feedback if you'd like some changes.

### CODE REVIEW

File: src/Loader.php [hunk 1]
```diff
@@ -236,7 +236,7 @@ protected function sanitiseVariableValue($name, $value)
                 %1$s           # match a quote at the start of the value
                 (              # capturing sub-pattern used
                  (?:           # we do not need to capture this
-                  [^%1$s\\\\]* # any character other than a quote or backslash
+                  [^%1$s\\\\]+ # any character other than a quote or backslash
```
Comment (GrahamCampbell) [inline src/Loader.php]: Why does this need to change please? Why can't we have the empty string?

File: src/Loader.php [hunk 2]
```diff
@@ -246,7 +246,19 @@ protected function sanitiseVariableValue($name, $value)
...
+            if(($e = preg_last_error()) !== PREG_NO_ERROR) {
+                $l = get_defined_constants(true);
+                $l = $l['pcre'];
+                $l = array_flip($l);
+                throw new InvalidFileException($l[$e]);
+            }
+
+            // if they are identical, it failed to replace/match
+            if($originalValue === $value) {
+                throw new InvalidFileException("Invalid value");
```
Comment (GrahamCampbell) [inline src/Loader.php]: I don't think this exception message is very clear?

File: src/Loader.php [hunk 3]
```diff
@@ -246,7 +246,19 @@ protected function sanitiseVariableValue($name, $value)
                 /mx',
                 $quote
             );
+            $originalValue = $value;
             $value = preg_replace($regexPattern, '$1', $value);
+            if(($e = preg_last_error()) !== PREG_NO_ERROR) {
+                $l = get_defined_constants(true);
+                $l = $l['pcre'];
+                $l = array_flip($l);
+                throw new InvalidFileException($l[$e]);
```
Comment (GrahamCampbell) [inline src/Loader.php]: I don't think the original error is going to make sense. Can we prepend some other message on the front?

File: src/Loader.php (hunk 1 above)
Comment (danielstokes) [reply src/Loader.php]: This doesn't need to change, just slightly better regex.  This forces a minimum of 1 characters for tokens not in (\ or (double)quote).  Line 242 ")*" is that actual matching token limit within the quoted value.

File: src/Loader.php (hunk 2 above)
Comment (danielstokes) [reply src/Loader.php]: I agree, I'll fix this.

File: src/Loader.php (hunk 3 above)
Comment (danielstokes) [reply src/Loader.php]: I think we could, I'll think of something.

**Comment by GrahamCampbell** [comment]
Ping @danielstokes. Are you able to finish this off please?

**Comment by danielstokes** [comment]
yeah, I can.  I'll finish it this week.

File: src/Loader.php [hunk 4]
```diff
@@ -246,7 +246,19 @@ protected function sanitiseVariableValue($name, $value)
                 /mx',
                 $quote
             );
+            $originalValue = $value;
             $value = preg_replace($regexPattern, '$1', $value);
+            if(($e = preg_last_error()) !== PREG_NO_ERROR) {
+                $l = get_defined_constants(true);
+                $l = $l['pcre'];
+                $l = array_flip($l);
```
Comment (GrahamCampbell) [inline src/Loader.php]: This won't work on PHP 7.3.

**Comment by GrahamCampbell** [comment]
I've been unable to come up with a test case that actually enters any of these cases. When can this possibly fail?

**Pull request closed without merging**